### PR TITLE
Fixes #1639 - Upgrade TS to 4.6.3

### LINKED ISSUE #1639: Dev: Update Project TS Version
- [ ] v3~ to v4.5

### PR DESCRIPTION
**Pull request by riotrah**
Update Project TS Version to 4.6.3 + some cspell
Fixes #1639

## What has Changed?

- Upgraded `typescript` version
- Upgraded `@typescript-eslint/eslint-plugin`
`@typescript-eslint/parser`
- Changed tsconfig lib entry to remove outdates EsNext.string entry
- Fixed some code that had changed semantics due to TS 4 changes to undefined/null etc

Fixes #1639 

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
     - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [x] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik

### CODE REVIEW

**Comment by riotrah** [comment]
I did a bunch of smoke testing via vsix in the ways I use it at work, but tbh this is a serious change and I _know_ I don't make use of all of Calva. I welcome advice, thorough feedback and testing please!

**Comment by bpringe** [comment]
I don't see anything that stands out to me, and the changes are few, really, but let's see what @PEZ and maybe @corasaurus-hex think (pinging her since she's done null check work recently).

File: src/edit.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export function continueCommentCommand() {
     }
     const commentOffset = cursor.rowCol[1];
     const commentText = cursor.getToken().raw;
-    const [_1, startText, bullet, num] = commentText.match(/^([;\s]+)([*-] +|(\d+)\. +)?/);
+    const [_1, startText, bullet, num] = commentText.match(/^([;\s]+)([*-] +|(\d+)\. +)?/) ?? [];
```
Comment (bpringe) [inline src/edit.ts:22]: This might deserve a look from @PEZ.

**Comment by corasaurus-hex** [comment]
Definitely merge this, this looks great!! I've wanted to do this for a long while but I prioritized null check work and keeping the pace of big changes to a manageable clip. This is the perfect time to merge this as I'll soon start on the null check work again and I'd love to have a newer (and smarter) TS to work from.

**Comment by corasaurus-hex** [comment]
Thanks so much for your contribution!! 💜

**Pull request merged**